src/workerd/api/worker-rpc.c++

```diff
@@ -2,19 +2,330 @@
+            builder.initValue().setV8Serialized(kj::mv(ok.serializedValue));
+          }
+          KJ_CASE_ONEOF(err, ThrewException) {
+            auto builder = callContext.initResults(capnp::MessageSize {
+                err.serializedException.size() / 8 + 8, 0 });
+            builder.initException().setV8Serialized(kj::mv(err.serializedException));
+          }
         }
-    ));
+      })));
+    });
+  }
+
+  KJ_DISALLOW_COPY_AND_MOVE(JsRpcTargetImpl);
+
+private:
+  // The following names are registered by the Workers Runtime and cannot be called over RPC.
+  bool isRegisteredName(kj::StringPtr name) {
+    if (name == "fetch"_kjc ||
+        name == "connect"_kjc ||
+        name == "alarm"_kjc ||
+        name == "webSocketMessage"_kjc ||
+        name == "webSocketClose"_kjc ||
+        name == "webSocketError"_kjc) {
+      return true;
+    }
+    return false;
+  }
+
+  // If the `methodName` is a known public method, we'll return it.
+  inline v8::Local<v8::Function> tryGetFn(
+      Worker::Lock& lock,
+      kj::Maybe<kj::StringPtr> entrypointName,
+      IoContext& ctx,
+      kj::StringPtr methodName) {
+
+    auto& handler = KJ_REQUIRE_NONNULL(lock.getExportedHandler(entrypointName, ctx.getActor()),
+        "Failed to get handler to worker.");
+
+    auto handle = handler.self.getHandle(lock);
+    auto methodStr = jsg::v8StrIntern(lock.getIsolate(), methodName);
+    auto fnHandle = jsg::check(handle->Get(lock.getContext(), methodStr));
+
+    jsg::Lock& js(lock);
+    v8::Local<v8::Object> obj = js.obj();
+    auto objProto = obj->GetPrototype().As<v8::Object>();
+
+    // Get() will check the Object and the prototype chain. We want to verify that the function
+    // we intend to call is not the one defined on the Object prototype.
+    bool isImplemented = jsg::check(handle->Get(js.v8Context(), methodStr)) !=
+                         jsg::check(objProto->Get(js.v8Context(), methodStr));
+
+    JSG_REQUIRE(isImplemented && fnHandle->IsFunction(), Error,
+        "The RPC receiver does not implement the requested method."_kjc);
+    JSG_REQUIRE(!isRegisteredName(methodName), Error,
+        kj::str("'", methodName, "' is a registered method and cannot be called over RPC."_kj));
+    return fnHandle.As<v8::Function>();
+  }
+
+  // Deserializes the arguments and passes them to the given function.
+  v8::Local<v8::Value> invokeFn(
+      jsg::Lock& js,
+      v8::Local<v8::Function> fn,
+      kj::ArrayPtr<const kj::byte> serializedArgs) {
+    // We received arguments from the client, deserialize them back to JS.
+    if (serializedArgs.size() > 0) {
+      auto args = KJ_REQUIRE_NONNULL(
+          deserializeV8(js, serializedArgs).tryCast<jsg::JsArray>(),
+          "expected JsArray when deserializing arguments.");
+      // Call() expects a `Local<Value> []`... so we populate an array.
+      auto arguments = kj::heapArray<v8::Local<v8::Value>>(args.size());
```

Comment:
Thanks, I was actually wondering about stack allocation here a while back but was put off because I assumed `args.size()` could be large: https://github.com/cloudflare/workerd/pull/1028/commits/7562375fb4e0c84af22da84426f7e3097dd4758f#diff-bbe685e15c44d4de0aed0b4b484f1c0fbe1427fccbf80ab9f65c6180f6632203R208